### PR TITLE
Clean up AbstractPost a bit

### PR DESCRIPTION
**Pull request by diegoreymendez**
This PR tries to clean up our codebase a bit by:

- Migrating some methods from AbstractPost to Swift.
- Removing some unused methods.

## Testing:

Make sure the unit tests run correctly.
Thoroughly review the code.

## Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

**Comment by jkmassel** [comment]
👋 I'm bumping this to `13.0` as part of the `12.9` code freeze. If you'd like it released as part of `12.9`, just merge it to `release/12.9` and DM me – I'll be happy to cut a new beta release!

**Comment by diegoreymendez** [comment]
@shiki - Adding you as a reviewer in case you have the bandwidth.  This is not really related to any project, but just something I do from time to time to clean up code and move a bit closer to having an all Swift codebase.

File: WordPress/Classes/Models/AbstractPost.swift [hunk 1]
```diff
@@ -1,10 +1,81 @@
...
+
+        return AbstractPost.title(for: status)
+    }
+
+    @objc
+    var remoteStatus: AbstractPostRemoteStatus {
+        get {
+            guard let remoteStatusNumber = remoteStatusNumber?.uintValue,
+                let status = AbstractPostRemoteStatus(rawValue: remoteStatusNumber) else {
+                    return .pushing
```
Comment (shiki) [inline WordPress/Classes/Models/AbstractPost.swift]: This is an unfortunate Swift conversion because we are required to provide a default value. But `.pushing` seems to be an incorrect default value for me. 

I do not think we should fix this though. My only concern here is, for someone new looking at this code, I don't quite understand the reason **why** `.pushing` is returned in the case that `remoteStatusNumber` is invalid. 

The `remoteStatusNumber` definition in the Core Data DB is also weird in that it has a default value, `0`, but is optional. ¯\\\_(ツ)\_/¯

File: WordPress/Classes/Models/AbstractPost.swift [hunk 2]
```diff
@@ -1,10 +1,81 @@
...
     class func title(for status: Status) -> String {
-        return AbstractPost.title(forStatus: status.rawValue)
+        return title(forStatus: status.rawValue)
+    }
+
+    /// Returns the localized title for the specified status.  Status should be
+    /// one of the `PostStatus...` constants.  If a matching title is not found
+    /// the status is returned.
+    ///
+    /// - parameter string: The post status value
```
Comment (shiki) [inline WordPress/Classes/Models/AbstractPost.swift]: This should be:

```swift
/// - parameter status: The post status value
```

🙂

File: WordPress/Classes/Models/AbstractPost.swift (hunk 1 above)
Comment (diegoreymendez) [reply WordPress/Classes/Models/AbstractPost.swift]: > This is an unfortunate Swift conversion because we are required to provide a default value. But .pushing seems to be an incorrect default value for me.
> 
> I do not think we should fix this though. My only concern here is, for someone new looking at this code, I don't quite understand the reason why .pushing is returned in the case that remoteStatusNumber is invalid.
> 
> The remoteStatusNumber definition in the Core Data DB is also weird in that it has a default value, 0, but is optional. ¯\_(ツ)_/¯

I agree.  And I also agree we shouldn't fix this for this PR, although I'd like to explain a bit more about "why".

The intent of this PR was to offer a migration that accurately maintains the existing logic.  This will allows us in part to minimize the intrinsic risk of the migration, and in part to scope what the extent of my work was for this PR making the review hopefully easier.

In a way, this migration is good because it highlights a shortcoming in the previously-implicit default value for this property that was sort of hidden before now.

File: WordPress/Classes/Models/AbstractPost.swift [hunk 3]
```diff
@@ -1,10 +1,81 @@
...
+            return NSLocalizedString("Trashed", comment: "Name for the status of a trashed post")
+        case PostStatusScheduled:
+            return NSLocalizedString("Scheduled", comment: "Name for the status of a scheduled post")
+        default:
+            return status
+        }
+    }
+
+    @objc
+    class func title(for remoteStatus: AbstractPostRemoteStatus) -> String {
```
[the remark this answers is not in the document]
Comment (diegoreymendez) [reply WordPress/Classes/Models/AbstractPost.swift]: Good catch.  I didn't realize this one was unused.  Removing.

File: WordPress/Classes/Models/AbstractPost.swift [hunk 4]
```diff
@@ -1,10 +1,81 @@
...
+
+    /// Returns the localized title for the specified status.  Status should be
+    /// one of the `PostStatus...` constants.  If a matching title is not found
+    /// the status is returned.
+    ///
+    /// - parameter string: The post status value
+    /// - returns: The localized title for the specified status, or the status if a title was not found.
+    ///
+    @objc
+    class func title(forStatus status: String) -> String {
```
[the remark this answers is not in the document]
Comment (diegoreymendez) [reply WordPress/Classes/Models/AbstractPost.swift]: Is there any specific reason why allowing these methods to be overridden is bad?  I'd rather make it more flexible unless there's a good reason not to.

File: WordPress/Classes/Models/AbstractPost.swift [hunk 5]
```diff
@@ -1,10 +1,81 @@
...
+        }
+
+        set {
+            remoteStatusNumber = NSNumber(value: newValue.rawValue)
+        }
+    }
+
     class func title(for status: Status) -> String {
-        return AbstractPost.title(forStatus: status.rawValue)
+        return title(forStatus: status.rawValue)
```
[the remark this answers is not in the document]
Comment (diegoreymendez) [reply WordPress/Classes/Models/AbstractPost.swift]: This is being called externally so it can't really be private.

Same question about `static` as above.

File: WordPress/Classes/Models/AbstractPost.swift (hunk 2 above)
Comment (diegoreymendez) [reply WordPress/Classes/Models/AbstractPost.swift]: Good observation.

**Comment by diegoreymendez** [comment]
@shiki - Ready for another look.  I haven't marked any of your observations as resolved, since I think it's best that you decide on it instead.

File: WordPress/Classes/Models/AbstractPost.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/Classes/Models/AbstractPost.swift]: Here are some questions that went into my head:

#### When would we ever want to override `title(forStatus:)`? 

I guess we would need to override this if `Post` and `Page` would consider `PostStatusDraft` and the other enum values as different. The chances of that ever happening are slim since `Post` and `Page` are mostly interchangeable. If it does happen, what does that mean for the call sites like these?

https://github.com/wordpress-mobile/WordPress-iOS/blob/8c42fc3c29ac9e4bf06318fd67780f8d9929e0bf/WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m#L1068-L1070

https://github.com/wordpress-mobile/WordPress-iOS/blob/f29e1679b2985145bb8a0b4037a11ef32bfc93c6/WordPress/Classes/Models/AbstractPost.swift#L8-L14

Does that mean them we'll have to change them to do type-checking to ensure that they're calling the correct overidden routines, `[Post titleForStatus:]` or `[Page titleForStatus:]`? That is quite cumbersome and I'd wager we're probably going to change `titleForStatus` if that happens. There are better solutions than overriding this method. 

#### How expensive is it to change in the future? 

Keeping it strictly `static` makes the next developer to touch this code to think twice before overidding the method. They will have to think if it really makes sense, probably considering my points above. If they do end up overidding, the change is not expensive. 

It is more expensive to keep it open and have the next developer ask questions like _"It is open for overrides. Are the subclasses overidding it? Why?"_ This adds additional debugging time for every single time you pass by this code. Or worse, they could override it and forget about how the call sites are using this method. Human mistakes are expensive. 

#### Summary

In summary, I believe it's just simply better to have a closed API rather than opening everything. It's harder to roll back when we've opened something to the public and we later realized that it was wrong to do that. 

Happy to hear your thoughts. 🙂

File: WordPress/Classes/Models/AbstractPost.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/Classes/Models/AbstractPost.swift]: That's weird. I changed this to `private` and the project still compiled. 

```swift 
    private static func title(for status: Status) -> String {
        return title(forStatus: status.rawValue)
    }
```

¯\\\_(ツ)\_/¯ 

Anyways, this is a special case. And I guess we do want it to be public since it's a utility function hiding inside a class.

**Comment by shiki** [comment]
@diegoreymendez I responded. Please let me know if I missed anything. 🙂

File: WordPress/Classes/Models/AbstractPost.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (diegoreymendez) [reply WordPress/Classes/Models/AbstractPost.swift]: By "this method" I imagine you refer to the method including the referenced line.  I assumed you meant the method called on the referenced line.

The whole idea of having this method to begin with is to promote the usage of the `Status` enum over the raw value string.  Making it private would indeed defeat that purpose.

[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/Classes/Models/AbstractPost.swift]: > Making it private would indeed defeat that purpose.

Yes, I agree. I was wrong about suggesting to make this private. It's probably why I mentioned I was not sure if it should be in https://github.com/wordpress-mobile/WordPress-iOS/pull/12106#discussion_r303956800.

[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/Classes/Models/AbstractPost.swift]: In the case of promoting the usage of the `Status` enum, I think we should place the `title` resolver in the `enum` itself.

```swift
    enum Status: String {
        case draft = "draft"
        case pending = "pending"
        case publishPrivate = "private"
        case publish = "publish"
        case scheduled = "future"
        case trash = "trash"
        case deleted = "deleted" // Returned by wpcom REST API when a post is permanently deleted.
        
        var title: String {
            switch self {
            case .draft:
                return NSLocalizedString("Draft", comment: "Name for the status of a draft post.")
            case .pending:
                return NSLocalizedString("Pending review", comment: "Name for the status of a post pending review.")
            case .publishPrivate:
                return NSLocalizedString("Private", comment: "Name for the status of a post that is marked private.")
            case .publish:
                return NSLocalizedString("Published", comment: "Name for the status of a published post.")
            case .trash:
                return NSLocalizedString("Trashed", comment: "Name for the status of a trashed post")
            case .scheduled:
                return NSLocalizedString("Scheduled", comment: "Name for the status of a scheduled post")
            default:
                return rawValue
            }
        }
    }
```

Not today, but probably in the future.

[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/Classes/Models/AbstractPost.swift]: > By "this method" I imagine you refer to the method including the referenced line. I assumed you meant the method called on the referenced line.

Yes, sorry about that. 😞

File: WordPress/Classes/Models/AbstractPost.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (diegoreymendez) [reply WordPress/Classes/Models/AbstractPost.swift]: > When would we ever want to override title(forStatus:)?

For example if a subclass wants to return a different title, such as "Draft Page" for pages.

The truth is that ultimately `AbstractPost` should most likely not be where this method lives.  The title returned by this method is "specific" to one or several VCs, but definitely not universal.

But I digress.

> How expensive is it to change in the future?

Honestly it depends on what you want to do in the future.  If you wanted to override the title for a specific subtype of post it'd argue it'd be less expensive to have this be `class` methods.  And if the developer sees this is `static` they'd be asking themselves "why" - which as you mention above is just as expensive as asking the same question the other way around.

The bottom line is that while I appreciate the effort to make a point for a possible improvement... I don't think this *should* be `static` any more than it should be `class` methods, and again - I prefer to allow these methods to be overriden since there's a chance the title may not be the same for a page and a post.

[the remark this answers is not in the document]
Comment (diegoreymendez) [reply WordPress/Classes/Models/AbstractPost.swift]: Actually, I'll just change this.

I still don't think this is strongly wrong, but I have no problem with this not being overridable for the time being.

File: WordPress/Classes/Models/AbstractPost.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (diegoreymendez) [reply WordPress/Classes/Models/AbstractPost.swift]: > In the case of promoting the usage of the Status enum, I think we should place the title resolver in the enum itself.

The scope of this PR is just to migrate the code, not to improve on it.  This is a deliberate choice  on my end to avoid coupling tasks that can be tackled separately.

[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/Classes/Models/AbstractPost.swift]: > The scope of this PR is just to migrate the code, not to improve on it. 

Right. I'm very sorry about that. I've gone way overboard on the review here. I didn't realize that you simply wanted to convert it to Swift without making any improvements since it adds to the development time. Again, my apologies. 😞

[the remark this answers is not in the document]
Comment (diegoreymendez) [reply WordPress/Classes/Models/AbstractPost.swift]: Hah, no worries it's fine... it's not about the time only though - the thing is more about separating concerns and keeping the PR scoped for me.

I've pushed some PRs in the past which suffered a bit from scope creep due to tackling "just one more thing", and now I'm trying to get into the mindset of being very mindful of the initial scope, and making it clear in the PR.

**Comment by diegoreymendez** [comment]
Thanks @shiki !  No worries!
